pkg/apis/seedmanagement/validation/managedseed.go

```diff
@@ -0,0 +1,327 @@
+	return allErrs
+}
+
+// ValidateManagedSeedUpdate validates a ManagedSeed object before an update.
+func ValidateManagedSeedUpdate(newManagedSeed, oldManagedSeed *seedmanagement.ManagedSeed) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&newManagedSeed.ObjectMeta, &oldManagedSeed.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, ValidateManagedSeedSpecUpdate(&newManagedSeed.Spec, &oldManagedSeed.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateManagedSeed(newManagedSeed)...)
+
+	return allErrs
+}
+
+// ValidateManagedSeedSpec validates the specification of a ManagedSeed object.
+func ValidateManagedSeedSpec(spec *seedmanagement.ManagedSeedSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Ensure shoot name is specified
+	if spec.Shoot.Name == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("shoot", "name"), "shoot name is required"))
+	}
+
+	// Ensure either seed template or gardenlet is specified
+	if (spec.SeedTemplate == nil && spec.Gardenlet == nil) || (spec.SeedTemplate != nil && spec.Gardenlet != nil) {
+		allErrs = append(allErrs, field.Invalid(fldPath, spec, "either seed template or gardenlet is required"))
+	}
+
+	switch {
+	case spec.SeedTemplate != nil:
+		allErrs = append(allErrs, validateSeedTemplate(spec.SeedTemplate, fldPath.Child("seedTemplate"))...)
+	case spec.Gardenlet != nil:
+		allErrs = append(allErrs, validateGardenlet(spec.Gardenlet, fldPath.Child("gardenlet"))...)
+	}
+
+	return allErrs
+}
+
+// ValidateManagedSeedSpecUpdate validates the specification updates of a ManagedSeed object.
+func ValidateManagedSeedSpecUpdate(newSpec, oldSpec *seedmanagement.ManagedSeedSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Ensure shoot name is not changed
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newSpec.Shoot.Name, oldSpec.Shoot.Name, fldPath.Child("shoot", "name"))...)
+
+	// Ensure no changing from seed template to gardenlet or vice versa takes place
+	if (newSpec.SeedTemplate != nil) != (oldSpec.SeedTemplate != nil) || (newSpec.Gardenlet != nil) != (oldSpec.Gardenlet != nil) {
+		allErrs = append(allErrs, field.Invalid(fldPath, newSpec, "changing from seed template to gardenlet and vice versa is not allowed"))
+	}
+
+	switch {
+	case newSpec.SeedTemplate != nil && oldSpec.SeedTemplate != nil:
+		allErrs = append(allErrs, validateSeedTemplateUpdate(newSpec.SeedTemplate, oldSpec.SeedTemplate, fldPath.Child("seedTemplate"))...)
+	case newSpec.Gardenlet != nil && oldSpec.Gardenlet != nil:
+		allErrs = append(allErrs, validateGardenletUpdate(newSpec.Gardenlet, oldSpec.Gardenlet, fldPath.Child("gardenlet"))...)
+	}
+
+	return allErrs
+}
+
+func validateSeedTemplate(seedTemplate *seedmanagement.SeedTemplate, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Validate labels, annotations, and spec
```

Comment:
I think, we should also enforce an empty `metadata.name` field in the seed template similar to the `gardenletConfig.seedConfig.metadata.name` check in the `validateSeed` function: https://github.com/gardener/gardener/blob/cde35895cd5fa3505e6970e41ba3ec2099e65d11/pkg/apis/seedmanagement/validation/managedseed.go#L282-L288

You can potentially even merge both funcs into one, which accepts a `metav1.ObjectMeta` and a `gardencore.SeedSpec`.

Reply:
Good point, I thought about it, I didn't do it because:
* I could easily add a check for the name - but what about other metadata fields, e.g. `namespace`, etc.? They will be ignored as well, shouldn't we also validate them?
* The Kubernetes deployment (and other similar objects) also silently ignores metadata fields different from labels and annotations, the validation doesn't fail if e.g. a name is specified.

The graceful way to handle this would be to not use `ObjectMeta` at all, but have a special type e.g. `RestrictedObjectMeta` that only has labels and annotations, so specifying a `Name` is not even possible. Didn't do it because:
* Again Kubernetes deployment etc. doesn't do it this way (though I believe it should).
* It would be inconsistent between `SeedTemplate` and `GardenletConfiguration`, and I was wary of changing `GardenletConfiguration`, besides the `Name` there is actually used by the `gardenlet`.

The cleanest (but involving the most changes and so potentially most time-consuming and dangerous) way to handle this IMO:
* Introduce `RestrictedObjectMeta` with labels and annotations only (perhaps in `core`).
* `SeedTemplate` uses `RestrictedObjectMeta` (could also be moved to `core`).
*  Both `ManagedSeed` and `GardenletConfiguration` use `SeedTemplate`. `GardenletConfiguration` has an additional `SeedName` field in `SeedConfig`.

What do you think? I am bit reluctant to change `GardenletConfiguration` at this point, on the other hand perhaps now is a better time than later.

Reply:
On the other hand, if we want to avoid such disruptive changes now (which I would prefer), then I could simply add the name validation as you suggested and be done with it.

Reply:
> On the other hand, if we want to avoid such disruptive changes now (which I would prefer), then we could simply add the name validation as you suggested and be done with it.

I would also keep the API types like they are (aligned with the k8s APIs) and only validate the `name` field.
I think, validating the other fields is not necessary, as they will be ignored. `name` will be overwritten by the controller and the reason for enforcing an unset `name` field, was to avoid confusion.
So, this is all good now from my perspective, thanks!